source/dub/generators/generator.d

```diff
@@ -143,130 +142,151 @@ class ProjectGenerator
+				tt = TargetType.none;
+			}
 
-		logDiagnostic("Generate target %s (%s %s %s)", pack.name, buildsettings.targetType, buildsettings.targetPath, buildsettings.targetName);
-		if (is_target)
-			targets[pack.name] = TargetInfo(pack, [pack], configs[pack.name], buildsettings, null);
+			setIs_target(tt);
+			shallowbs.targetType = tt;
 
-		auto deps = pack.getDependencies(configs[pack.name]);
-		foreach (depname; deps.keys.sort()) {
-			auto depspec = deps[depname];
-			auto dep = m_project.getDependency(depname, depspec.optional);
-			if (!dep) continue;
+			if (tt == TargetType.none) {
+				// ignore any build settings for targetType none (only dependencies will be processed)
+				shallowbs = BuildSettings.init;
+				shallowbs.targetType = TargetType.none;
+			}
 
-			auto depbs = collect(settings, dep, targets, configs, main_files, is_target ? pack.name : bin_pack);
+			// start to build up the build settings
+			processVars(buildsettings, m_project, pack, shallowbs, true);
 
-			if (depbs.targetType != TargetType.sourceLibrary && depbs.targetType != TargetType.none) {
-				// add a reference to the target binary and remove all source files in the dependency build settings
-				depbs.sourceFiles = depbs.sourceFiles.filter!(f => f.isLinkerFile()).array;
-				depbs.importFiles = null;
+			// remove any mainSourceFile from library builds
+			if (buildsettings.targetType != TargetType.executable && buildsettings.mainSourceFile.length) {
+				buildsettings.sourceFiles = buildsettings.sourceFiles.filter!(f => f != buildsettings.mainSourceFile)().array;
+				main_files ~= buildsettings.mainSourceFile;
 			}
 
-			buildsettings.add(depbs);
+			// set pic for dynamic library builds.
+			if (buildsettings.targetType == TargetType.dynamicLibrary)
+				buildsettings.addOptions(BuildOption.pic);
 
-			if (depbs.targetType == TargetType.executable)
-				continue;
+			logDiagnostic("Generate target %s (%s %s %s)", pack.name, buildsettings.targetType, buildsettings.targetPath, buildsettings.targetName);
+			if (is_target)
+				targets[pack.name] = TargetInfo(pack, [pack], configs[pack.name], buildsettings, null);
 
-			auto pt = (is_target ? pack.name : bin_pack) in targets;
-			assert(pt !is null);
-			if (auto pdt = depname in targets) {
-				pt.dependencies ~= depname;
-				pt.linkDependencies ~= depname;
-				if (depbs.targetType == TargetType.staticLibrary)
-					pt.linkDependencies = pt.linkDependencies.filter!(d => !pdt.linkDependencies.canFind(d)).array ~ pdt.linkDependencies;
-			} else pt.packages ~= dep;
+			forDependenciesBuildSettings = buildsettings.dup;
 		}
+		else {
+			buildsettings = packInTargets.buildSettings;
+			forDependenciesBuildSettings = buildsettings.dup;
+			setIs_target(buildsettings.targetType);
+		}
+		mergeFromDependents(parentSettings, forDependenciesBuildSettings);
 
-		if (is_target) targets[pack.name].buildSettings = buildsettings.dup;
+		if (auto p = pack.name in forDependenciesSettingsMap)
```

Comment:
What's the point of this `forDependenciesSettingsMap`? Are you caching data flows from top to bottom, so it can be reused for multiple children?
I guess we're doing a depth-first-order traversal, why isn't it enough to dup before recursing? Diamond dependencies?

Reply:
Diamonds. Packages get visited more than once, must remember and re-use results from the previous visit(s).

Reply:
I don't see a reason why versions shouldn't simply always be additive. That's how they are designed in the language and the library can easily check for conflicting semantics. IMO it would unnecessarily complicate DUB to start assigning semantics on that level.

Reply:
What I meant is *being used additively*, so EnableFeatureX instead of DisableFeatureX, that way conflicts in diamond dependencies should be less of a problem.
Indeed I wouldn't want dub to have anything to do with this, but it might be a best practice for library developers. But even more I'd suggest people to use configurations as a cleaner mechanism instead.